### ISSUE TITLE
manual section that explains the flow of the program

### ISSUE THREAD

**Issue by jdannberg** [geodynamics/aspect#2282]
This is something that came up in a discussion at the CGU/CIG meeting. 
Several participants mentioned that they would find it useful to get some more documentation on the core of ASPECT, namely, what happens when the program is started and what are the main functions that are called from there. This should probably include a flowchart that links the flow of the program to the files and functions where things are happening. 

Traditionally, this is something that we do not expose very much, but people who have used other codes before find it useful to see where the time stepping/assembly/etc. is done.

**Comment by bangerth** [comment]
The problem is how to keep this kind of document up to date. We do change things in the code base, and it's easy to forget that a certain section in the manual needs to be updated as well. Maybe we can limit this to the big picture, rather than technical details?

**Comment by gassmoeller** [comment]
We should not put in a lot of details that require frequent updates, but many participants in our tutorial found it very valuable to hear about where the program starts, how it flows, and what is called from where. I think steps that should be included in the chart would be:
- position of `int main()` and that it creates a `Simulator`, and then calls `Simulator.run()`
- position of `Simulator` and that `run()` contains the main time loop, and calls functions like `start_timestep`,`solve_timestep`, `postprocessing`, etc.
- position of the solver schemes, solver functions, and assembly functions
- some information about how our plugin system relates to the directory structure

These are not things that are of interested to participants of a very basic tutorial (like students who just started geodynamic modelling), but we had a large crowd of people developing other codes, and they are right, that we are not doing a good job of explaining how the internals of ASPECT work.

**Comment by tjhei** [comment]
duplicate of #1029 ? The plugin graph there might be a good place to start...

**Comment by gassmoeller** [comment]
@tjhei: Maybe, but we were explicitly asked: Where is this code? Not: How does this conceptually work?

**Comment by tjhei** [comment]
I disagree. If you look at my figure, it is already a flow graph. If you annotate this with class/function names and .cc files, you would be quite close to what I would do. I think having the places where plugins are "active" in the graph is very useful.

this is a sketch for what I would do:
```
main() -- see main in main.cc

  construct Simulator -- ::Simulator()

    create geometry  -> plugin: GeometryModel

    set_assemblers   -> plugin: ...


  Simulator::run() -- run() in simulator/core.cc

    initial adaptive refinement loop:
      set_initial_temperature_and_compositional_fields -> plugin: initial conditions
      maybe_refine_mesh

    loop:
      start_timestep()
        ...

      solve_timestep()
        ...

      postprocessing -- ::postprocess()  -> pl

      maybe_refine_mesh
```

Does that make sense?

**Comment by bangerth** [comment]
I like @tjhei 's flow chart/pseudocode. 

That said, it really only gives a very abbreviated view of things. Much more happens -- for example, reading parameters, creating all of the other plugin systems, etc. I wonder how much detail is necessary.

@gassmoeller -- what specifically did they want to know? Were they just asking where all of the ASPECT code is, or did they specifically ask about individual pieces? What was the context of their questions?
